engine/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/MostSpecificValueExtractorTest.java

```diff
@@ -0,0 +1,139 @@
+
+import org.hibernate.validator.HibernateValidator;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.Entity1;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.Entity2;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper11;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper111;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper21;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper211;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper212;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper22;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.IWrapper221;
+import org.hibernate.validator.test.internal.engine.valuehandling.model.Wrapper2;
+import org.hibernate.validator.testutil.TestForIssue;
+import org.testng.annotations.Test;
+
+/**
+ * @author Guillaume Smet
+ */
+public class MostSpecificValueExtractorTest {
+
+	@Test
+	@TestForIssue(jiraKey = "HV-1306")
+	public void mostSpecificValueExtractorFound() throws Exception {
+		Validator validator = Validation.byProvider( HibernateValidator.class )
+				.configure()
+				.addValueExtractor( new IWrapper11ValueExtractor() )
+				.addValueExtractor( new IWrapper111ValueExtractor() )
+				.buildValidatorFactory()
+				.getValidator();
+
+		Set<ConstraintViolation<Entity1>> violations = validator.validate( new Entity1( null ) );
+		assertCorrectPropertyPaths( violations, "wrapper.IWrapper11" );
+	}
+
+	@Test
+	@TestForIssue(jiraKey = "HV-1306")
+	public void parallelValueExtractorDefinitionsCausesException() throws Exception {
+		Validator validator = Validation.byProvider( HibernateValidator.class )
+				.configure()
+				.addValueExtractor( new IWrapper21ValueExtractor() )
+				.addValueExtractor( new IWrapper211ValueExtractor() )
+				.addValueExtractor( new IWrapper212ValueExtractor() )
+				.addValueExtractor( new IWrapper22ValueExtractor() )
+				.addValueExtractor( new IWrapper221ValueExtractor() )
+				.buildValidatorFactory()
+				.getValidator();
+
+		try {
+			validator.validate( new Entity2( null ) );
+			fail( "An exception should have been thrown" );
+		}
+		catch (ConstraintDeclarationException e) {
+			String message = e.getMessage();
+			assertThat( message ).startsWith( "HV000219" );
+			assertThat( message ).contains( Wrapper2.class.getName() );
+			assertThat( message ).contains( "IWrapper21ValueExtractor" );
+			assertThat( message ).doesNotContain( "IWrapper211ValueExtractor" );
+			assertThat( message ).doesNotContain( "IWrapper212ValueExtractor" );
+			assertThat( message ).contains( "IWrapper22ValueExtractor" );
+			assertThat( message ).doesNotContain( "IWrapper221ValueExtractor" );
+		}
+	}
+
+	private static class IWrapper11ValueExtractor implements ValueExtractor<IWrapper11<@ExtractedValue ?>> {
```

Comment:
Can we try and use it Lambda-style (inline when calling `addValueExtractor()`), too? Just to make sure that this works.